### DISCUSSION TITLE
Kotlin optional parameter changed in 4.0.0?

### DISCUSSION THREAD

**Discussion by jarlehansen** [ExpediaGroup/graphql-kotlin#1126] [General] [answered]
We had this code working in version 3.x
```
 suspend fun vaults(names: List<String>?)
```

Where names is set to `null` if no parameter was set in the query. However, when upgrading to version 4.0.0 I started to get this error message:
```
No argument provided for a required parameter: parameter #2 names of ...
``` 

It is fixed by setting `names: List<String>? = null`, but is this how it is intended to work?

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
Yes it was intentional. In older versions we always defaulted to null for optional arguments which while convenient was also ignoring any non-null defaults set in the code, i.e. given `fun foo(bar: String? = "baz")` it would always default to `bar = null`. While we cannot read those default values using reflections (so cannot expose them in the schema), we changed the default data fetcher logic to only set the argument value if it was specified in the request so it correctly honors the defaults now - in the example below `bar` will default to `baz`. As a side effect this requires all optional arguments to specify their defaults explicitly.

See https://github.com/ExpediaGroup/graphql-kotlin/pull/981

**Reply by jarlehansen** [reply]
thanks for the answer, sounds good to me. Always missed the ability to specify default values, so good to see it implemented now :)
